File: neoforge/src/main/java/org/bread_experts_group/breadlib/NeoEvents.java
package org.bread_experts_group.breadlib;

import net.minecraft.client.Minecraft;
import net.minecraft.server.level.ServerPlayer;
import net.neoforged.bus.api.Event;
import net.neoforged.bus.api.IEventBus;
import net.neoforged.neoforge.client.event.ClientTickEvent;
import net.neoforged.neoforge.client.event.InputEvent;
import net.neoforged.neoforge.client.event.RegisterGuiLayersEvent;
import net.neoforged.neoforge.client.event.RenderLevelStageEvent;
import net.neoforged.neoforge.client.gui.VanillaGuiLayers;
import net.neoforged.neoforge.common.NeoForge;
import net.neoforged.neoforge.event.tick.ServerTickEvent;
import net.neoforged.neoforge.network.event.RegisterPayloadHandlersEvent;
import net.neoforged.neoforge.network.registration.PayloadRegistrar;
import org.bread_experts_group.breadlib.network.context.ServerNetworkContext;
import org.bread_experts_group.breadlib.network.payload.PayloadInfo;
import org.bread_experts_group.breadlib.task.FireSide;
import org.bread_experts_group.breadlib.task.TaskManager;
import org.bread_experts_group.breadlib.task.input.MouseTasks;
import org.bread_experts_group.breadlib.task.network.NetworkTask;
import org.bread_experts_group.breadlib.task.render.LayeredDrawTask;
import org.bread_experts_group.breadlib.task.render.LevelRenderTask;
import org.bread_experts_group.breadlib.task.render.RenderLevelStage;
import org.bread_experts_group.breadlib.task.tick.ClientTickTask;
import org.bread_experts_group.breadlib.task.tick.ServerTickTask;

import java.util.function.Consumer;

import static org.bread_experts_group.breadlib.task.render.RenderLevelStage.*;

public class NeoEvents {
	private static <T extends Event> void addListener(Class<T> eventClass, Consumer<T> task) {
		NeoForge.EVENT_BUS.addListener(eventClass, task);
	}

	private static RenderLevelStage getRenderLevelStage(RenderLevelStageEvent.Stage stage) {
		if (stage == RenderLevelStageEvent.Stage.AFTER_SKY) {
			return AFTER_SKY;
		} else if (stage == RenderLevelStageEvent.Stage.AFTER_SOLID_BLOCKS) {
			return AFTER_SOLID_BLOCKS;
		} else if (stage == RenderLevelStageEvent.Stage.AFTER_CUTOUT_MIPPED_BLOCKS_BLOCKS) {
			return AFTER_CUTOUT_MIPPED_BLOCKS_BLOCKS;
		} else if (stage == RenderLevelStageEvent.Stage.AFTER_CUTOUT_BLOCKS) {
			return AFTER_CUTOUT_BLOCKS;
		} else if (stage == RenderLevelStageEvent.Stage.AFTER_ENTITIES) {
			return AFTER_ENTITIES;
		} else if (stage == RenderLevelStageEvent.Stage.AFTER_BLOCK_ENTITIES) {
			return AFTER_BLOCK_ENTITIES;
		} else if (stage == RenderLevelStageEvent.Stage.AFTER_TRANSLUCENT_BLOCKS) {
			return AFTER_TRANSLUCENT_BLOCKS;
		} else if (stage == RenderLevelStageEvent.Stage.AFTER_TRIPWIRE_BLOCKS) {
			return AFTER_TRIPWIRE_BLOCKS;
		} else if (stage == RenderLevelStageEvent.Stage.AFTER_PARTICLES) {
			return AFTER_PARTICLES;
		} else if (stage == RenderLevelStageEvent.Stage.AFTER_WEATHER) {
			return AFTER_WEATHER;
		} else if (stage == RenderLevelStageEvent.Stage.AFTER_LEVEL) {
			return AFTER_LEVEL;
		}

		throw new ClassCastException("Failed to map NeoForge specific RenderLevelStageEvent.Stage: " + stage);
	}

	public static void registerEvents(IEventBus eventBus) {
		addRLSETask();
		addMouseScrollTask();
		addMouseButtonTasks();
		addClientTickTasks();
		addServerTickTasks();
		addLayeredDrawTask(eventBus);
		addNetworkTasks(eventBus);
	}

	private static void addRLSETask() {
		addListener(RenderLevelStageEvent.class, (event) -> TaskManager.runTasks(
				new LevelRenderTask(
						getRenderLevelStage(event.getStage()),
						event.getLevelRenderer(),
						event.getPoseStack(),
						event.getProjectionMatrix(),
						event.getPartialTick(),
						event.getCamera(),
						event.getFrustum()
				))
		);
	}

	private static void addMouseScrollTask() {
		addListener(InputEvent.MouseScrollingEvent.class, event -> {
			if (TaskManager.runTasks(
					new MouseTasks.Scroll(Minecraft.getInstance().mouseHandler, event.getScrollDeltaX(), event.getScrollDeltaY())
			).isCanceled()) event.setCanceled(true);
		});
	}

	private static void addMouseButtonTasks() {
		addListener(InputEvent.MouseButton.Pre.class, event -> {
			if (TaskManager.runTasks(
					new MouseTasks.Button(
							Minecraft.getInstance().mouseHandler,
							event.getButton(),
							event.getAction(),
							event.getModifiers(),
							FireSide.PRE
					)
			).isCanceled()) event.setCanceled(true);
		});
		addListener(InputEvent.MouseButton.Post.class, event -> TaskManager.runTasks(
				new MouseTasks.Button(
						Minecraft.getInstance().mouseHandler,
						event.getButton(),
						event.getAction(),
						event.getModifiers(),
						FireSide.POST
				)
		));
	}

	private static void addClientTickTasks() {
		addListener(ClientTickEvent.Pre.class, event -> TaskManager.runTasks(
				new ClientTickTask(Minecraft.getInstance().level, FireSide.PRE)
		));
		addListener(ClientTickEvent.Post.class, event -> TaskManager.runTasks(
				new ClientTickTask(Minecraft.getInstance().level, FireSide.POST)
		));
	}

	private static void addServerTickTasks() {
		addListener(ServerTickEvent.Pre.class, event -> TaskManager.runTasks(
				new ServerTickTask(event.getServer().overworld(), FireSide.PRE)
		));
		addListener(ServerTickEvent.Post.class, event -> TaskManager.runTasks(
				new ServerTickTask(event.getServer().overworld(), FireSide.POST)
		));
	}

	private static void addLayeredDrawTask(IEventBus eventBus) {
		eventBus.addListener(RegisterGuiLayersEvent.class, event -> {
			LayeredDrawTask task = TaskManager.runTasks(new LayeredDrawTask());
			task.getLayers().forEach((location, layer) ->
					event.registerAbove(VanillaGuiLayers.DEBUG_OVERLAY, location, layer)
			);
		});
	}

	@SuppressWarnings({"unchecked", "rawtypes"})
	private static void addNetworkTasks(IEventBus eventBus) {
		eventBus.addListener(RegisterPayloadHandlersEvent.class, event -> {
			NetworkTask task = TaskManager.runTasks(new NetworkTask());
			PayloadRegistrar registrar = event.registrar("1.4.0");

			for (PayloadInfo info : task.serverboundPayloads()) {
				registrar.playToServer(
						info.type(),
						info.streamCodec(),
						(payload, context) -> {
							info.handler().handle(payload, new ServerNetworkContext((ServerPlayer) context.player()));
				});
			}
		});
	}
}

File: neoforge/src/main/kotlin/org/bread_experts_group/breadlib/NeoEvents.kt
package org.bread_experts_group.breadlib

import net.minecraft.client.Minecraft
import net.minecraft.client.gui.LayeredDraw
import net.minecraft.resources.ResourceLocation
import net.neoforged.bus.api.Event
import net.neoforged.bus.api.IEventBus
import net.neoforged.neoforge.client.event.*
import net.neoforged.neoforge.client.gui.VanillaGuiLayers
import net.neoforged.neoforge.common.NeoForge
import net.neoforged.neoforge.event.RegisterCommandsEvent
import net.neoforged.neoforge.event.tick.ServerTickEvent
import org.bread_experts_group.breadlib.task.FireSide
import org.bread_experts_group.breadlib.task.TaskManager
import org.bread_experts_group.breadlib.task.command.ClientCommandTask
import org.bread_experts_group.breadlib.task.command.ServerCommandTask
import org.bread_experts_group.breadlib.task.input.KeyboardTask
import org.bread_experts_group.breadlib.task.input.MouseTasks
import org.bread_experts_group.breadlib.task.render.LayeredDrawTask
import org.bread_experts_group.breadlib.task.render.LevelRenderTask
import org.bread_experts_group.breadlib.task.render.RenderLevelStage
import org.bread_experts_group.breadlib.task.tick.ClientTickTask
import org.bread_experts_group.breadlib.task.tick.ServerTickTask
import java.util.function.Consumer

object NeoEvents {
	private fun <T : Event> addListener(task: Consumer<T>): Unit = NeoForge.EVENT_BUS.addListener(task)

	private fun getRenderLevelStage(stage: RenderLevelStageEvent.Stage): RenderLevelStage =
		when (stage) {
			RenderLevelStageEvent.Stage.AFTER_SKY -> RenderLevelStage.AFTER_SKY
			RenderLevelStageEvent.Stage.AFTER_SOLID_BLOCKS -> RenderLevelStage.AFTER_SOLID_BLOCKS
			RenderLevelStageEvent.Stage.AFTER_CUTOUT_MIPPED_BLOCKS_BLOCKS -> RenderLevelStage.AFTER_CUTOUT_MIPPED_BLOCKS_BLOCKS
			RenderLevelStageEvent.Stage.AFTER_CUTOUT_BLOCKS -> RenderLevelStage.AFTER_CUTOUT_BLOCKS
			RenderLevelStageEvent.Stage.AFTER_ENTITIES -> RenderLevelStage.AFTER_ENTITIES
			RenderLevelStageEvent.Stage.AFTER_BLOCK_ENTITIES -> RenderLevelStage.AFTER_BLOCK_ENTITIES
			RenderLevelStageEvent.Stage.AFTER_TRANSLUCENT_BLOCKS -> RenderLevelStage.AFTER_TRANSLUCENT_BLOCKS
			RenderLevelStageEvent.Stage.AFTER_TRIPWIRE_BLOCKS -> RenderLevelStage.AFTER_TRIPWIRE_BLOCKS
			RenderLevelStageEvent.Stage.AFTER_PARTICLES -> RenderLevelStage.AFTER_PARTICLES
			RenderLevelStageEvent.Stage.AFTER_WEATHER -> RenderLevelStage.AFTER_WEATHER
			RenderLevelStageEvent.Stage.AFTER_LEVEL -> RenderLevelStage.AFTER_LEVEL
			else -> throw IllegalStateException("Failed to map NeoForge specific RenderLevelStageEvent.Stage: $stage")
		}

	fun registerEvents(eventBus: IEventBus) {
		this.addRLSETask()
		this.addMouseScrollTask()
		this.addMouseButtonTasks()
		this.addKeyboardTasks()
		this.addClientTickTasks()
		this.addServerTickTasks()
		this.addLayeredDrawTask(eventBus)
		this.addCommandTasks()
	}

	private fun addRLSETask() {
		this.addListener { event: RenderLevelStageEvent ->
			TaskManager.runTasks(
				LevelRenderTask(
					this.getRenderLevelStage(event.stage),
					event.levelRenderer,
					event.poseStack,
					event.projectionMatrix,
					event.partialTick,
					event.camera,
					event.frustum
				)
			)
		}
	}

	private fun addMouseScrollTask() {
		this.addListener { event: InputEvent.MouseScrollingEvent ->
			if (TaskManager.runTasks(
					MouseTasks.Scroll(
						Minecraft.getInstance().mouseHandler,
						event.scrollDeltaX,
						event.scrollDeltaY
					)
				).isCanceled
			) event.setCanceled(true)
		}
	}

	private fun addMouseButtonTasks() {
		this.addListener { event: InputEvent.MouseButton.Pre ->
			if (TaskManager.runTasks(
					MouseTasks.Button(
						Minecraft.getInstance().mouseHandler,
						event.button,
						event.action,
						event.modifiers,
						FireSide.PRE
					)
				).isCanceled
			) event.setCanceled(true)
		}
		this.addListener { event: InputEvent.MouseButton.Post ->
			TaskManager.runTasks(
				MouseTasks.Button(
					Minecraft.getInstance().mouseHandler,
					event.button,
					event.action,
					event.modifiers,
					FireSide.POST
				)
			)
		}
	}

	private fun addKeyboardTasks() {
		this.addListener { event: InputEvent.Key ->
			TaskManager.runTasks(
				KeyboardTask(event.key, event.scanCode, event.action, event.modifiers)
			)
		}
	}

	private fun addClientTickTasks() {
		val level = Minecraft.getInstance().level ?: return
		this.addListener { _: ClientTickEvent.Pre ->
			TaskManager.runTasks(ClientTickTask(level, FireSide.PRE))
		}
		this.addListener { _: ClientTickEvent.Post ->
			TaskManager.runTasks(ClientTickTask(level, FireSide.POST))
		}
	}

	private fun addServerTickTasks() {
		this.addListener { event: ServerTickEvent.Pre ->
			TaskManager.runTasks(ServerTickTask(event.server.overworld(), FireSide.PRE))
		}
		this.addListener { event: ServerTickEvent.Post ->
			TaskManager.runTasks(ServerTickTask(event.server.overworld(), FireSide.POST))
		}
	}

	private fun addLayeredDrawTask(eventBus: IEventBus) {
		eventBus.addListener { event: RegisterGuiLayersEvent ->
			val task = TaskManager.runTasks(LayeredDrawTask())
			task.layers.forEach { (location: ResourceLocation, layer: LayeredDraw.Layer) ->
				event.registerAbove(VanillaGuiLayers.DEBUG_OVERLAY, location, layer)
			}
		}
	}

	private fun addCommandTasks() {
		this.addListener { event: RegisterClientCommandsEvent ->
			TaskManager.runTasks(ClientCommandTask(event.dispatcher, event.buildContext))
		}
		this.addListener { event: RegisterCommandsEvent ->
			TaskManager.runTasks(ServerCommandTask(event.dispatcher, event.buildContext))
		}
	}
}